### ISSUE TITLE
Timber sends an excessive amound of DELETES to the database

### ISSUE THREAD

**Issue by dennisjac** [timber/timber#1834]
### Expected behavior
Timber should not send an excessive number of requests to the database.

### Actual behavior
Timber sends an excessive amount of DELETE queries to the database server in an attempt to purge transients.

### Steps to reproduce behavior
Call the render() function often with a non-false value for the $expires parameter.

### What version of WordPress, PHP and Timber are you using?
Timber 1.7.1, Wordpress 4.9.8, PHP 7.0

### How did you install Timber? (for example, from GitHub, Composer/Packagist, WP.org?)
Packagist

### Notes
In `lib/Loader.php::render()` the `delete_cache()` function is called every time the function is called with a non-false $expires parameter which in our setup causes an excessive number of DELETE queries that make up a non-trivial amount of total queries sent to the database (this DELETE query is sent several times each second).

The core Wordpress code already does an expiration and it uses a scheduled event on order to prevent this kind of flood. See `wp-admin/admin.php`:
```
if ( ! wp_next_scheduled( 'delete_expired_transients' ) && ! wp_installing() ) {
        wp_schedule_event( time(), 'daily', 'delete_expired_transients' );
}
```
Timber should adopt a similar approach to cap the amount of cache purges to a more reasonable value.

**Comment by chads2000** [comment]
Just hit this issue, it basically took down website down for a few hours while it was clearing.

I have also noticed that this system does a hard delete of the cache, essentially removing and inserting a new.  For a site that is ~5 Years old, my wp_options primary index is getting over 160mil.  Wondering if would be better to instead of removing the record, but to null the value until the next write.

**Comment by chads2000** [comment]
Can we get an update or movement on this. Here is my DB connection stack right now. Over 126 connections!!!!!

<img width="605" alt="Screen Shot 2020-04-17 at 11 17 43 AM" src="https://user-images.githubusercontent.com/1414604/79596196-5efbc300-809d-11ea-92bc-bfa901c7a0eb.png">

**Comment by jarednova** [comment]
hi @chads2000 — I'm turning my attention to this now. Thanks for your follow-ups; this is helpful to see if I can suss out what's going on here

**Comment by chads2000** [comment]
This may help

calling `Timber::render('index.twig', $data, 600);` with an expiration time.

It is passed to the `self::fetch($filenames, $data, $expires, $cache_mode);` function and on to `self::compile($filenames, $data, $expires, $cache_mode, true);` and on to `$loader->render($file, $data, $expires, $cache_mode);`

Inside `Loader::render` this if block exists

```
if (false !== $output && false !== $expires && null !== $key) {
  $this->delete_cache();
  $this->set_cache($key, $output, self::CACHEGROUP, $expires, $cache_mode);
}
```

Since `$expires` doesn't equal `false` as it is an int value, this if block is always called, and the `$this->delete_cache()` function is called sending Deletes to the database.

But since ` $this->set_cache` uses the WP functions to set transients and caching and the `$expires` value is passed along with it.  Wouldn't the initial call to `$this->get_cache($key, self::CACHEGROUP, $cache_mode)` result in there being a cached object or not, and there for we don't need to "DELETE" the cache?

The other is `$this->delete_cache()` has no prejudice in cache clearing, and should be reduced to just clearing out the cache it is trying to set by tracking the cache `$key`.

**Comment by jarednova** [comment]
@chads2000 Thanks for tracking that down. I'm wondering if we stick this patch in at the top of `Loader::render` ...

```php
if ( $expires === 0 ) {
	$expires = false;
}
```

... does that cure the error for you?

**Comment by chads2000** [comment]
Yeah that should patch the issue.

**Comment by parkerwebteam** [comment]
Looks to be some positive discussions being had regarding this potential issue. Cheers @chads200 and @jarednova. 

Personally we have found on a production level that our servers have spiked in a similar way, and have *sometimes* gone down for a period of time. Of course, every project is different with other complex differences, but we did arrive at this issue and thought this could be a potential cause, and of course we can't be 100% sure.  

However, after a server down, we investigated our logs on & before the period and discovered the below in our logs:

`
WordPress database error Deadlock found when trying to get lock; try restarting transaction for query UPDATE `jgk_options` SET `option_value` = '2194015847' WHERE `option_name` = '_transient_timeout_timberloader_f794af87e2a0b0f83a5f97e2ae354760' made by require('wp-blog-header.php'), require_once('wp-includes/template-loader.php'), include('/themes/xxxxxx/template-home.php'), Timber\\Timber::render, Timber\\Timber::fetch, Timber\\Timber::compile, Timber\\Loader->render, Timber\\Loader->set_cache, set_transient, update_option
`

As above, we are caching a file with the default cache type using:

```
Timber::render('template-home.twig', $context, 604800000);
```

With `604800000` being one week.

We just thought we'd mention this as it could be useful to know that more than one party is having a similar issue.

It could be entirely coincidental that the above is related to transient/database errors, but worth mentioning i'm sure.

**Comment by jarednova** [comment]
@chads2000 @parkerwebteam thanks for the follow-ups. I just merged #2243 to resolve this for now. Though, on a larger task we should consider removing "built-in" caching functionality and separating into an easily documented or controlled module for the future.

I'm cl

**Comment by countfak** [comment]
Hi,

I'm using Timber 2.0.0-alpha.4 and I'm having this issue.
Is it solved on the stable 1.x branch? Should it be resolved on the 2.x branch too? Would using the 2.x-dev release help?

Thanks
